src/mlmc/sample_storage.py

```diff
@@ -0,0 +1,86 @@
+from abc import ABCMeta
+from abc import abstractmethod
+from typing import List
+import re
+from new_simulation import QuantitySpec
+from workspace import Workspace
+
+
+class SampleStorage(metaclass=ABCMeta):
+
+    @abstractmethod
+    def save_results(self, res):
+        """
+        Write results to storag
+        """
+
+    @abstractmethod
+    def save_result_format(self, res_spec: List[QuantitySpec]):
+        """
+        Save result format
+        """
+
+    @abstractmethod
+    def load_result_format(self) -> List[QuantitySpec]:
+        """
+        Load result format
+        """
+
+    @abstractmethod
+    def save_workspace(self, workspace: Workspace):
+        """
+        Save some workspace attributes
+        :return: None
+        """
+
+    @abstractmethod
+    def sample_pairs(self):
+        """
+        Get results from storage
+        :return:
+        """
+
+    def _get_level(self, sample_id: str):
+        return re.findall(r'L0?(\d+)_', sample_id)[0]
+
+
+class Memory(SampleStorage):
+
+    def __init__(self):
+        self._results = {}
+        self._scheduled = {}
+        self._result_specification = []
+
+    def save_results(self, results):
```

Comment:
Sampler can keep samples from individual levels separated and pass them separated to the storage. So results, as well as self._results would be list of tables one for every level.